### ISSUE TITLE
Disconnect handler not executed if connection drops during middleware execution

### ISSUE THREAD

**Issue by adroste** [socketio/socket.io#4129]
**Describe the bug**
If I use a middleware (e.g. for authentication) I cannot handle a client disconnect because
1. `disconnect` event-handler does not get executed
2. `socket.connected` won't update properly (from `true` to `false`)

**To Reproduce**

Please fill the following code example:

Socket.IO server version: `4.2.0`

*Server*

```js
import { Server } from "socket.io";

const io = new Server(3000, {});

io.use(async (socket, next) => {
    console.log(`middleware`, socket.id);

    // bug 1: disconnect handler will never execute
    socket.on("disconnect", () => {
      console.log(`disconnect ${socket.id}`);
    });
    
    // long running async operation
    // client will disconnect before operation has finished
    await new Promise(resolve => setTimeout(resolve, 3000));

    // bug 2: connected state is still true
    console.log(`still connected: ${socket.connected}`);

    next();
});

io.on("connection", (socket) => {
    console.log(`connect ${socket.id}`);
});
```

Socket.IO client version: `4`

*Client*

```js
import { io } from "socket.io-client";

const socket = io("ws://localhost:3000/", {});
// force disconnect before long running operation has finished
setTimeout(window.location.reload, 1000);
```

**Expected behavior**
`socket.connected` will update properly and `disconnect` event-handler will run

**Platform:**
node.js 16 on macOS

**Additional context**
The workaround to the problem is: don't trigger side-effects in middleware and always add custom properties to the socket object to transfer information over to `connection` event-handler

**Comment by kanhaiya-2000** [comment]
@adroste Middleware is called before the socket connection is established. You can move your `disconnect` event logic into the block
 `io.on("connection",(socket)=>{
   //Rest code above
    socket.on("disconnect", () => {
       console.log(`disconnect ${socket.id}`);
    });
    
})` and it should work.

**Comment by darrachequesne** [comment]
@adroste this behavior is expected. The `socket` is not considered connected when the middleware gets executed, it will once the last middleware of the chain calls `next()`.

**Comment by darrachequesne** [comment]
> Nevertheless, I still think that the socket.connected property should be properly updated and set to false on early disconnect or is not set to true before the "connection" event is fired.

This was fixed in https://github.com/socketio/socket.io/commit/02b0f73e2c64b09c72c5fbf7dc5f059557bdbe50, included in version [4.4.0](https://github.com/socketio/socket.io/releases/tag/4.4.0).

Please reopen if needed.
